### PR TITLE
Feat/md2 tabs

### LINKED ISSUE #3617: [Bug Report] v-tabs is not updating lazyValue properly
### Versions and Environment
**Vuetify:** 1.0.8
**Vue:** 2.5.16
**Browsers:** Chrome 65.0.3325.162
**OS:** Windows 7

### Steps to reproduce
- Open the reproduction link https://codepen.io/anon/pen/XEpmrR
- No tab should be active, but "Pricing" is active;
- Navigate to "/about";
- The "About" tab is active as expected;
- Navigate to "/other";
- No tab should be active, but "About" is active.

### Expected Behavior
- No tab should be active as default (before routing);
- No tab should be active when in other route than "/pricing" or "/about".

### Actual Behavior
The last tab stays active when the current route doesn't match to any tab `:to`

### Reproduction Link
[https://codepen.io/anon/pen/XEpmrR](https://codepen.io/anon/pen/XEpmrR)

### Other comments
The behavior I expect for tabs is that a tab is active only when it is on its route (or its children).

### LINKED ISSUE #3631: [Bug Report] Tabs (Android): Slider animation is delayed after swipe
### Versions and Environment
**Vuetify:** 1.0.9
**Vue:** 2.5.7
**Browsers:** Google Chrome 65.0.3325.109
**OS:** Android 5.1.1

### Steps to reproduce
1. Go to https://vuetifyjs.com/en/components/tabs using Chrome@Android
2. Swipe between tabs in example

### Expected Behavior
Slider moves with tab transition (like when you click on a tab)

### Actual Behavior
Slider animation is delayed and plays after tab change
http://s.csssr.ru/U07B23NE8/screencast_2018-03-20_19-00-24.mp4
(video captured from actual device using Chrome's inspect)

### Reproduction Link
https://vuetifyjs.com/en/components/tabs

### LINKED ISSUE #3633: [Feature Request] Add a hover effect to v-tabs
### Problem to solve
Desktop users can know better if they are hovering a tab.

### Proposed solution
Add a hover behavior on tabs:
http://material-components-web.appspot.com/tabs.html

v-tabs can have a `hover` prop (default true).

### LINKED ISSUE #3971: [Feature Request] v-tabs Vertical option
### Problem to solve
i need use other vue plugin to create vertical tabs in complex projects 

### Proposed solution
implement option to use

### LINKED ISSUE #4204: [Bug Report] Support large lists of tabs on sections of a page (i.e. doesn't take up the whole width of the screen and still needs to be scrollable)
### Versions and Environment
**Vuetify:** 1.0.17
**Vue:** 2.5.16
**Browsers:** Google Chrome
**OS:** Mac OS 10.13.4

### Steps to reproduce
1. Click on right most tab

### Expected Behavior
It should scroll across and show the next tab (i.e. tabs should go up to 'news 10')

### Actual Behavior
No scrolling

### Reproduction Link
[https://codepen.io/chluke/pen/Wyrpxa](https://codepen.io/chluke/pen/Wyrpxa)

### Other comments
Currently the scrolling of the tabs seems to work best when it occupies the full width of the screen. There are occasions when you might only use part of the screen. For example, on mobiles you could have a clickable navigation drawer, but on PC's you have it permanently shown - so the tabs in this chase only occupy: Screen width - 300px (drawer). This means that the "tab-wrapper" transform: translateX(-XXXXpx); is now always 300px off.

### LINKED ISSUE #4420: [Bug Report] tabs have extra margins when in a toolbar
### Versions and Environment
**Vuetify:** 1.1.0-rc.3
**Vue:** 2.5.16
**Browsers:** Chrome 67.0.3396.87
**OS:** Windows 10

### Steps to reproduce
Resize to ~1300px width

### Expected Behavior
As in 1.0, tabs should be right against the edge: https://codepen.io/anon/pen/qKLNoM?editors=1000

### Actual Behavior
There is an extra 24px margin on either side

### Reproduction Link
[https://codepen.io/anon/pen/yEGJEX?editors=1000](https://codepen.io/anon/pen/yEGJEX?editors=1000)

### Other comments
#4305 should be reverted, gutters were complex for a reason. 
This can't be easily fixed with negative margins. 

Mobile: 
image

Arrows visible:
image

### LINKED ISSUE #4735: [Bug Report] Setting active tab after removing an tab from array causes selection offset
### Versions and Environment
Vuetify: 1.1.8
Vue: 2.5.16
Browsers: Chrome v68
OS: Windows 10

### Steps to reproduce

Click on an X button to close a tab. (as noted below, closing the last tab functions properly)

### Expected Behavior

The tab should close and the last tab should be selected.

### Actual Behavior

The slider bar selects the correct tab However the active tab is the 1 index behind. When clicking the tab with the slider, it behaves as if it was the active tab (nothing happens).

### Reproduction Link

[https://codepen.io/MajesticPotatoe/pen/rrGoEG](https://codepen.io/MajesticPotatoe/pen/rrGoEG)

---

### Additional Comments:
This behavior does not happen when iterating through active tabs (like in docs) or if you remove the last tab. Have also tried doing async/await but everything seems to debug as if its functioning normally.

One thing to note that may need a new ticket is that if you remove all of the tabs from the array, the slider bar still remains.

### LINKED ISSUE #5398: [Bug Report] New v-tab does not scroll into view with show-arrows
When I add more tabs that can fit in the width of the tabs bar, and the scroll arrows show, the first new tab does not scroll into view even though I set the v-model to it's index.

### Versions and Environment
Vuetify: 1.3.1
Vue: 2.5.17
Browsers: 69.0.3497.100
OS: MacOS High Sierra 10.13.6

### Reproduction Link

Here is a link to a repro: https://codepen.io/pbastowski/pen/pxOONM?editors=1100

### Steps to reproduce

If you keep clicking the ADD button at the end of the tabs bar you will notice that tab 5 is hidden, see image below.

image

If you add another tab, tab 6 on my screen, then both tabs 5 and 6 are fully revealed.

image

I have a workaround for this right now, see the code below. `tabs` is an array, which I use to render each `v-tab`. When the array's length increases I scroll the new tab into view using `scrollIntoView()`, but only after 700ms have passed, otherwise it does not work (it could be something to do with the animation, I guess).

```js
    watch: {
        /*
            Fix for the hidden new tab problem. 
            Always scroll the last added tab into view
         */
        'tabs.length'(nv, ov) {
            if (!this.$refs.tabs) return  // $refs doe not exist until mounted
            setTimeout(() => {
                if (nv === ov + 1)
                    try {
                        this.$refs.tabs.$el
                            .querySelector('.s-tab:last-child')
                            .scrollIntoView(false)
                    } catch (e) {
                        // silently ignore
                    }
            }, 700)
        }
    },
```

### Expected Behavior

The new tab scrolls into view automatically when I set the `v-model` on `v-tabs` to it's index, even when it's the first new tab that does not fit into the available width of the tabs bar.

### Actual Behavior

The newly added tab remains hidden. Adding another new tab, scrolls both new tabs into view, see images above.

### LINKED ISSUE #5515: [Bug Report] Deleting a tab in v-tabs causes errors
### Versions and Environment
**Vuetify:** 1.3.5
**Vue:** 2.5.17
**Browsers:** Chrome 70.0.3538.77
**OS:** Windows 10

### Steps to reproduce
# Deleting tab(s) will cause different errors
Case A shows what happens out of the box, while Case B is my attempt to change the active tab before it get's removed, which helps a bit, but there are still errors left.

## Case A) Without manually changing "active"
("Preset active" checkbox unchecked)

**Bug 1:**
- Steps: Add 3 tabs, press "DEL TAB 1"
- Result (Bug): The 3rd tab is removed and 2nd is activated, but "active" stays on "2", and the slider stays on the 3rd entry

**Bug 2:**
- Steps: Add 3 tabs, select the first one, press "DEL TAB 0", and then try to change tabs
- Result (Bug): Slider will not move anymore, and when clicking a tab, it will select the other one instead

## Case B) Manually changing "active" before removing a tab
("Preset active" checkbox checked)

**Bug3:**
- Steps: Add 3 tabs, select the first one, press "DEL TAB 1", and then try to change tabs
- Result (Bug): Tabs won't react to clicks anymore

### Expected Behavior
When removing tab(s), "active" should be recalculated and set automatically (see Chrome browser tabs as an example which tab should be the next active one)

### Actual Behavior
After tab(s) are being removed, "active" is wrong, the slider will be at the wrong place, and tab changes will no longer work or have unexpected behavior (select wrong one etc)

### Reproduction Link
<a href="https://codepen.io/anon/pen/yQBEmz?editors=1010" target="_blank">https://codepen.io/anon/pen/yQBEmz?editors=1010</a>

### LINKED ISSUE #5762: [Bug Report] VTab to and VTabItem combination crashing browser tab
### Versions and Environment
**Vuetify:** 1.3.11
**Vue:** 2.5.17
**Browsers:** Chrome 70.0.3498.0, Safari, Mozilla Firefox
**OS:** Mac OS 10.13.4

### Steps to reproduce
The exact point of failure is having v-tabs and a v-tab-item and then adding a `to` attribute to one of the `v-tab`s

```
<v-tabs>
  <v-tab to="/test">tab1</v-tab>
  <v-tab>tab2</v-tab>
  <v-tab>tab3</v-tab>

  <v-tab-item></v-tab-item>
</v-tabs>
```

### Expected Behavior
The browser should not crash. I'd like to be able to use both a `to` and the `v-tab-item` so that users can visit tabs via direct link and use the tabs as usual 

### Actual Behavior
The tab freezes and becomes unresponsive. 

### Reproduction Link
<a href="https://codepen.io/anon/pen/QJJLNg" target="_blank">https://codepen.io/anon/pen/QJJLNg</a>

### Other comments
There may be a better way of doing my use case, but the fact it's crashing the browser isn't good.

### LINKED ISSUE #6014: [Bug Report] v-item-group/v-item incorrectly tracks active v-item
### Versions and Environment
**Vuetify:** 1.3.15
**Vue:** 2.5.21
**Browsers:** Firefox 64.0, Google Chrome
**OS:** Windows 7, Linux

### Steps to reproduce
Click the plus icon until you have three tabs. The last tab should be shown as active.
Click the minus icon. The first tab should be active. Instead the last tab is incorrectly being shown as active

### Expected Behavior
Click the minus icon. The first tab should be active.

### Actual Behavior
Click the minus icon. The last tab is incorrectly being shown as active

### Reproduction Link
<a href="https://codepen.io/uglyhobbitfeet/pen/xmPjqM?editors=1111" target="_blank">https://codepen.io/uglyhobbitfeet/pen/xmPjqM?editors=1111</a>

### LINKED ISSUE #6278: [Bug Report] Assigning new array in v-tabs component causes issues
### Versions and Environment
**Vuetify:** 1.4.3
**Vue:** 2.5.22
**Browsers:** Chrome 71.0.3578.98
**OS:** Mac OS 10.12.4

### Steps to reproduce
(1) Create a v-tab component and a v-tab-item component, both with v-for directives iterating over a data property "items" and a key set to "item.id"
(2) In the data method, define "items" to be an array of objects with a property "id."
(3) At some point after the initial render, assign a new array to the reactive data property "items" with a subset of the original array.

### Expected Behavior
The user should be able to activate all of the tabs by clicking on them.

### Actual Behavior
Not all of the tabs are able to be activated.

### Reproduction Link
<a href="https://codepen.io/corbatacolombiana/pen/NoraNK" target="_blank">https://codepen.io/corbatacolombiana/pen/NoraNK</a>

### LINKED ISSUE #6505: [Bug Report] v-tab does not deactivate when route changes to a non-tabbed route
### Versions and Environment
**Vuetify:** 1.5.1
**Vue:** 2.6.4
**Browsers:** Chrome 72.0.3626.109
**OS:** Mac OS 10.13.6

### Steps to reproduce
Using vue-router and the `:to` parameter.  Create a route in your app that is not available via v-tab.  Navigate to it and see that the previously selected route/tab is still showing as active in the v-tab

### Expected Behavior
De-activate tab if route does not match the tab.

### Actual Behavior
Previously selected tab stays active even when route changes.

### Reproduction Link
<a href="https://codepen.io/lucille2/pen/bzMmPQ" target="_blank">https://codepen.io/lucille2/pen/bzMmPQ</a>

### Other comments
In my app I use a v-toolbar header with tabs in the extension slot and an account menu in the main slot. The account menu can go to routes that are not available as tabs. But I cannot de-activate the v-tab when the user goes to one of those other routes.  **Using vue-router

### LINKED ISSUE #6527: [Bug Report] V-tab: Conditional tab messes up tabs on change
### Versions and Environment
**Vuetify:** 1.5.1
**Vue:** 2.6.6
**Browsers:** Chrome 72.0.3626.109
**OS:** Windows 10

### Steps to reproduce
Create a v-tab with named keys, set one tab and tab-item to conditional with v-if. Changing the v-if variable removes the tab and the tab-item, but you cannot shwitch to some of the tabs that should be still visible and accessible. 

### Expected Behavior
Clicking on the tab should take you to the tab-item even if other tabs are removed.

### Actual Behavior
If a tab (and tab-item) is removed by swicthing v-if, some of the tabs are not responding, tab-content does not change.

### Reproduction Link
<a href="https://codepen.io/t-th-tam-s/pen/JxwjNz" target="_blank">https://codepen.io/t-th-tam-s/pen/JxwjNz</a>

### Other comments
It might relate to named "key".

### LINKED ISSUE #6575: [Bug Report] VTabs pagination int rtl mode wont work properly
### Versions and Environment
**Vuetify:** 1.5.2
**Vue:** 2.6.6
**Browsers:** Chrome 72.0.3626.109
**OS:** Windows 10

### Steps to reproduce
clone repository.
run `npm install`
run `npm run serve`

### Expected Behavior
when the `<v-tabs>` component is overflowed and when `$vuetify.rtl` option is set to `true`, the items shall be in their right place.

### Actual Behavior
the `<v-tab>` in `<v-tabs>` seem to be off and not to be in their right places when the `<v-tabs>` overflows, and pagination will show a big offset to the right.

### Reproduction Link
<a href="https://github.com/FunnyPhantom/vuetify-bug-report" target="_blank">https://github.com/FunnyPhantom/vuetify-bug-report</a>

### PR DESCRIPTION
**Pull request by johnleider**
## Description
Converted `v-tabs` to typescript. Updated to match md2 spec styling.

fixes #3617
fixes #3631
fixes #3633 
resolves #3971
fixes #4204
fixes #4420
fixes #4735
fixes #5398
fixes #5515
fixes #5762 
fixes #6014
fixes #6278
fixes #6505 
fixes #6527
fixes #6575

## Motivation and Context
Update tabs to md2 spec and implement QoL fixes/changes.

## How Has This Been Tested?
jest

## Markup:
Use Playground or Docs page
</details>

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/components/VWindow/VWindowItem.ts [hunk 1]
```diff
@@ -56,6 +57,9 @@ export default mixins<options & ExtractVue<[typeof Bootable]>>(
   },
 
   computed: {
+    classes (): object {
+      return this.groupClasses
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VWindow/VWindowItem.ts]: what's the point of computed here?

File: packages/vuetify/src/components/VTabs/VTabsBar.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    $route: 'onRouteChange'
+  },
+
+  methods: {
+    callSlider () {
+      if (!this.isBooted) return
+
+      this.$emit('call:slider')
+    },
+    onRouteChange (val: any, oldVal: any) {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VTabs/VTabsBar.ts]:
```suggestion
    onRouteChange (val: Route, oldVal: Route) {
```

File: packages/vuetify/src/components/VTabs/VTabs.ts [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      }
+    }
+  },
+
+  watch: {
+    alignWithTitle: 'callSlider',
+    centered: 'callSlider',
+    fixedTabs: 'callSlider',
+    right: 'callSlider',
+    vertical: 'callSlider',
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VTabs/VTabs.ts]: also `showArrows` and `grow`

File: packages/vuetify/src/components/VTabs/VTab.ts [hunk 4]
```diff
@@ -86,18 +98,25 @@ export default {
...
+    data.attrs = {
+      ...data.attrs,
+      'aria-selected': String(this.isActive),
+      role: 'tab',
+      tabindex: 0
+    }
+    data.on = {
+      ...data.on,
+      keydown: (e: KeyboardEvent) => {
+        if (e.keyCode === 13) this.click(e)
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VTabs/VTab.ts]: `keyCodes.enter`

File: packages/vuetify/src/components/VWindow/VWindowItem.ts (hunk 1 above)
Comment (johnleider) [reply packages/vuetify/src/components/VWindow/VWindowItem.ts]: Helps with typings. The `v-item-group` classes doesn't make it through.

**Comment by DRoet** [comment]
Just FYI the `desktop tabs` and `vertical-tabs` seem broken in `IE11`. rest looks good! 👍

**Comment by johnleider** [comment]
> Just FYI the `desktop tabs` and `vertical-tabs` seem broken in `IE11`. rest looks good! 👍

Ty! Will square away before Tuesday

File: packages/vuetify/src/components/VTabs/VTabs.ts [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+  data () {
+    return {
+      resizeTimeout: 0,
+      slider: {
+        height: null as null | number,
+        left: null as null | number,
+        top: null as null | number,
+        width: null as null | number
+      } as Record<string, number | null>,
```
Comment (nekosaur) [inline packages/vuetify/src/components/VTabs/VTabs.ts]: This makes the typing of `slider` looser than it has to be.

File: packages/vuetify/src/components/VTabs/VTabs.ts [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+            this.internalValue = val
+          }
+        },
+        ref: 'items'
+      }), [
+        this.genSlider(slider),
+        items
+      ])
+    },
+    genItems (items: VNode[], item: VNode[]) {
```
Comment (nekosaur) [inline packages/vuetify/src/components/VTabs/VTabs.ts]: This was a confusing method and list or arguments. Perhaps a comment is warranted, or better argument names?

File: packages/vuetify/src/components/VTabs/VTabs.ts [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+      }, items)
+    },
+    onResize () {
+      if (this._isDestroyed) return
+
+      clearTimeout(this.resizeTimeout)
+      this.resizeTimeout = window.setTimeout(this.callSlider, 0)
+    },
+    parseNodes () {
+      const item = []
```
Comment (nekosaur) [inline packages/vuetify/src/components/VTabs/VTabs.ts]: Same as method above, having both item and items be arrays is counter-intuitive.

**Pull request merged**